lib/Clam/CfgBuilder.cc

```diff
@@ -2636,6 +2726,37 @@ void CrabIntraBlockBuilder::visitLoadInst(LoadInst &I) {
       } else if (rgn.getRegionInfo().getType().second > lhs_bitwidth) {
         m_bb.truncate(lhs_v, ext_or_trunc_lhs_v);
       }
+
+      auto it = m_ref_fat_map.find(I.getPointerOperand());
+      if (it == m_ref_fat_map.end()) {
+        CLAM_WARNING("Could not find load ptr from map "
+                     << *I.getPointerOperand());
+      } else {
+        // check dereference is in bounds
+        // ref on operand:
+        // assert(offset >= 0)
+        // assert(obj size > offset)
+        CRAB_LOG("cfg-fat", llvm::errs()
+                                << "load: check size and offset are in bounds: "
+                                << I << "\n");
+        var_t ptr_obj_sz = it->second.first;
+        var_t ptr_offset = it->second.second;
+        // uint64_t tmp = lhs_bitwidth / 8;
+        m_bb.assertion(ptr_offset >= number_t(0), getDebugLoc(&I));
+        m_bb.assertion(ptr_obj_sz > ptr_offset, getDebugLoc(&I));
+        if (lhs->getVar().get_type().is_reference()) {
+          var_t ref_obj_sz = m_lfac.mkIntVar(lhs_bitwidth);
+          // FIXME: obtain an overappriximation of deref
+          m_bb.assign(ref_obj_sz, number_t(0));
```

Comment:
this is not an over-approximation. You should do `havoc(ref_obj_sz)` and `havoc(ref_offset)`. Maybe we want to have some user options to give some values.